### PR TITLE
Turn `Connection` into a trait

### PR DESCRIPTION
**Pull request by sgrif**
This was pretty straightforward. Almost everything that was previously
coupled to PG has been changed to be generic over the connection (mainly
`LoadDsl` and `ExecuteDsl`). Our tests are coupled to PG. We'll want to
change that as we further break up PG into it's own crate. Things that
apply to Sqlite will need to be tested there as well.

`SimpleConnection` exists because I didn't want to make migrations
generic over the connection, but `Connection` is not object safe. We
only ever use `batch_execute` in that module, so I'm just taking that as
a trait object.

CLI and Codegen are currently coupled to PG. I will need to write
additional code for both of them in order to support SQLite in the
future. It occurs to me that I don't have a great way to make these
support third party adapters, and I'll need to rethink how the
connection info is passed in.

### CODE REVIEW

**Comment by sgrif** [comment]
@mfpiccolo @mcasper Can I get a quick review on this?

File: diesel_codegen/src/update.rs [hunk 1]
```diff
@@ -138,14 +138,15 @@ fn save_changes_impl(
         let pk_field = pk.field_name.unwrap();
         quote_item!(cx,
             impl<'a> $struct_name {
-                $_pub fn save_changes<T>(&self, connection: &::diesel::Connection)
+                $_pub fn save_changes<T, Conn>(&self, connection: &Conn)
```
Comment (mcasper) [inline diesel_codegen/src/update.rs]: Why does this also return the connection now?

Comment (mcasper) [inline diesel_codegen/src/update.rs]: Scratch that, misread it

Comment (sgrif) [inline diesel_codegen/src/update.rs]: This has the unfortunate side effect of causing `save_changes::<User>(&connection)` to turn into `save_changes::<User, _>(&connection)`. I'll fix this before the 0.5 release

Comment (sgrif) [inline diesel_codegen/src/update.rs]: Just need to pull this up to a `SaveChangesDsl<Conn: Connection>` and impl that instead.

**Comment by mcasper** [comment]
Looks good

**Pull request merged**